pytorch/pytorch
# [pytorch][mobile] add NO_EXPORT macro to unset __visibility__ attribute

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25896 [pytorch] update build_android.sh to not build host protoc for libtorch
* #25894 [pytorch] remove pthreadpool dependency in aten/CMake
* #25817 [pytorch][mobile] disable backward function codegen if disable_autograd is set
* **#25816 [pytorch][mobile] add NO_EXPORT macro to unset __visibility__ attribute**
* #25697 [pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile
* #25815 [pytorch][mobile] gate static aten registerer with USE_STATIC_DISPATCH

Summary:
On Android we will release a small set of native APIs designed for mobile use
cases. All of needed libtorch c++ APIs are called from inside this JNI bridge:
android/pytorch_android/src/main/cpp/pytorch_jni.cpp

With NO_EXPORT set for android static library build, it will hide all
original TORCH, CAFFE2, TH/ATen APIs, which will allow linker to strip
out unused ones from mobile library when producing DSO.

If people choose to directly build libtorch DSO then it will still keep
all c++ APIs as the mobile API layer is not part of libtorch build (yet).

Test Plan:
- build libtorch statically and link into demo app;
- confirm that linker can strip out unused APIs;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25816

Differential Revision: [D17247237](https://our.internmc.facebook.com/intern/diff/D17247237)
